backend/app/modules/multisig_signatures/service/offchain.py

```diff
@@ -48,15 +62,27 @@ def approve_pending_signatures(self, _: Context) -> list[Signature]:
             )
 
             if len(confirmations) >= threshold:
-                pending_signature.status = SigStatus.APPROVED
-                db.session.commit()
-
+                self.staged_signatures.append(pending_signature)
                 approved_signatures.append(
-                    Signature(pending_signature.message, pending_signature.hash)
+                    Signature(
+                        pending_signature.id,
+                        pending_signature.message,
+                        pending_signature.hash,
+                    )
                 )
 
+        # db.session.commit()
+
         return approved_signatures
 
+    def apply_staged_signatures(self, signature_id: int):
+        for idx, pending_signature in enumerate(self.staged_signatures):
+            if pending_signature.id == signature_id:
+                pending_signature.status = SigStatus.APPROVED
+                db.session.commit()
```

Comment:
in fact we do commit the pending signatures separately but it helps us to keep the consistency within doing actions after each approval (e.g triggering allocate() that may fail for each one of signatures)